lib/madness/command_line.rb

```diff
@@ -127,5 +132,49 @@ def build_toc
     def config
       @config ||= Settings.instance
     end
+
+    # Returns the URL where madness is running, suitable for opening in the
+    # browser
+    def server_url
+      scheme = ENV['MADNESS_FORCE_SSL'] ? 'https' : 'http'
+      host = config.bind == '0.0.0.0' ? 'localhost' : config.bind
+      port = config.port
+      
+      "#{scheme}://#{host}:#{port}"
+    end
+
+    # Returns truthy if the server is running. Will attempt to connect 
+    # multiple times. This is designed to assisn in running some code after
+    # the server has launched.
+    def server_running?(retries: 5, delay: 1)
+      connected = false
+      attempts = 0
+
+      begin
+        connected = Socket.tcp(config.bind, config.port)
+      rescue
+        sleep delay
+        retry if (attempts += 1) < retries
+      end
+
+      connected
+    end
+
+    # Open a web browser if the server is running. This is done in a
+    # non-blocking manner, so it can be executed before starting the server.
+    def open_browser
+      fork { open_browser! if server_running? }
```

Comment:
Tiny note: when server startup takes > 5.x seconds, browser will not be opened and no message will be displayed on STDOUT.

Reply:
Nice catch. Fixed.